k4Reco/GaudiLumiCalClusterer/components/GaudiLumiCalClusterer.cpp

```diff
@@ -0,0 +1,123 @@
+
+  LumiCalClusterer.createDecoder(fieldDescription);
+
+  // Set parameters for the GlobalMethodsClass
+  std::map<std::string, std::variant<int, float, std::string>> parameters;
+  parameters["ZLayerPhiOffset"] = static_cast<float>(m_zLayerPhiOffset);
+  parameters["EnergyCalibConst"] = static_cast<float>(m_EnergyCalibConst);
+  parameters["LogWeigthConstant"] = static_cast<float>(m_logWeigthConstant);
+  parameters["MinHitEnergy"] = static_cast<float>(m_minHitEnergy);
+  parameters["MiddleEnergyHitBoundFrac"] = static_cast<float>(m_MiddleEnergyHitBoundFrac);
+  parameters["ElementsPercentInShowerPeakLayer"] = static_cast<float>(m_ElementsPercentInShowerPeakLayer);
+  parameters["ClusterMinNumHits"] = static_cast<int>(m_ClusterMinNumHits);
+  parameters["MoliereRadius"] = static_cast<float>(m_rMoliere);
+  parameters["MinClusterEngy"] = static_cast<float>(m_minClusterEngy);
+  parameters["WeightingMethod"] = m_WeightingMethod;
+  parameters["NumOfNearNeighbor"] = static_cast<int>(m_NumOfNearNeighbor);
+
+  gmc.SetConstants(parameters);
+
+  m_BeamCrossingAngle = gmc.GlobalParamD[GlobalMethodsClass::BeamCrossingAngle] / 2.;
+
+  // printParameters();
+  // /* --------------------------------------------------------------------------
+  //    Print out Processor Parameters
+  //    -------------------------------------------------------------------------- */
+  // info() << "Global parameters for Processor:" << name() << "\t" << type() << endmsg;
+  // gmc.PrintAllParameters();
+  // info() << endmsg;
+
+  LumiCalClusterer.init(gmc);
+  LumiCalClusterer.setCutOnFiducialVolume(m_cutOnFiducialVolume);
+
+  return StatusCode::SUCCESS;
+}
+
+std::tuple<edm4hep::CalorimeterHitCollection, edm4hep::ClusterCollection, edm4hep::ReconstructedParticleCollection>
+GaudiLumiCalClusterer::operator()(const edm4hep::SimCalorimeterHitCollection& input) const {
+  /* --------------------------------------------------------------------------
+     create clusters using: LumiCalClustererClass
+     -------------------------------------------------------------------------- */
+  // TODO: Remove const_cast
+  auto [status, calhits] = LumiCalClusterer.processEvent(const_cast<edm4hep::SimCalorimeterHitCollection&>(input));
+
+  auto LCalClusterCol = edm4hep::ClusterCollection();
+
+  auto LCalRPCol = edm4hep::ReconstructedParticleCollection();
+
+  if (status == RETVAL::OK) {
+    debug() << " Transfering reco results to LCalClusterCollection....." << endmsg;
+
+    for (const int armNow : {-1, 1}) {
+      const auto& pairIDCellsVector = LumiCalClusterer.m_superClusterIdToCellId.at(armNow);
+      debug() << " Arm  " << std::setw(4) << armNow << "\t Number of clusters: " << pairIDCellsVector.size() << endmsg;
+
+      for (const auto& pairIDCells : pairIDCellsVector) {
+        const int clusterId = pairIDCells.first;
+        LCCluster& thisClusterInfo =
+            const_cast<LCCluster&>(LumiCalClusterer.m_superClusterIdClusterInfo.at(armNow).at(clusterId));
```

Comment:
Can the API be adapted to not have `const_cast`s?

Reply:
Not while `LumiCalClusterer` keeps internal state, which makes this not be ready for multithreading.